lib/spack/spack/cmd/uninstall.py

```diff
@@ -136,6 +158,12 @@ def do_uninstall(specs, force):
             # want to uninstall.
             spack.package.Package.uninstall_by_spec(item, force=True)
 
+        if ev.active:
+            try:
+                ev.active.remove(item, force=True)
+            except ev.EnvError:
+                pass  # ignore errors from specs that are not roots
```

Comment:
(clarification requested) The comment is tough for me to understand: `Environment.remove` complains if the specified Spec isn't in the environment, it seems like that should always be an error. Also it looks like `cmd.uninstall.find_matching_specs` will already complain if a Spec doesn't exist in an environment.

Reply:
`add` and `remove` just deal with user specs (roots) in `spack.yaml`, but you can still force-uninstall their dependencies.  I agree this isn't so clear and I don't really like squashing the exception. Would you rather it only remove if the spec is in `ev.active.user_specs`?  That seems cleaner to me, but I don't know if that is more understandable.

Reply:
The comment should be clearer: I think the comment should explain what is being done in the `try` block rather than why nothing is wrong in the `except` block

```
# try removing the spec from the current active environment. this will fail if the
# spec is not a user/root spec, which is ignored
```

Also, I'm concerned about automatically removing specs from environments when uninstalling. It's not necessarily wrong, but it may need careful documenting:

* How does this interact with reinstallation logic (`spack install --overwrite`)? In the overwrite case this spec is not removed, and that may be preferable. I think it's worth mentioning in the docs though
* That this occurs automatically for the current environment (which may be active implicitly through `SPACK_ENV`) could be problematic.

In short, there's some power to keeping the environment edit functions and installation management logic separate, which this undoes. My preference would be that auto-removing the spec would be an option for uninstall rather than automatic.

The main thing is the comment though.